sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/changefeed/implementation/ChangeFeedProcessorBuilderImpl.java

```diff
@@ -209,6 +214,92 @@ public boolean isStarted() {
             });
     }
 
+    /**
+     * Returns a list of states each representing one scoped worker item.
+     * <p>
+     * An empty list will be returned if the processor was not started or no lease items matching the current
+     *   {@link ChangeFeedProcessor} instance's lease prefix could be found.
+     *
+     * @return a list of states each representing one scoped worker item.
+     */
+    @Override
+    public Mono<List<ChangeFeedProcessorState>> getCurrentState() {
+        List<ChangeFeedProcessorState> earlyResult = new ArrayList<>();
+
+        if (this.leaseStoreManager == null || this.feedContextClient == null) {
+            return Mono.just(earlyResult);
+        }
+
+        return this.leaseStoreManager.getAllLeases()
+            .flatMap(lease -> {
+                ChangeFeedOptions options = new ChangeFeedOptions()
+                    .setMaxItemCount(1)
+                    .setPartitionKeyRangeId(lease.getLeaseToken())
+                    .setStartFromBeginning(true)
+                    .setRequestContinuation(lease.getContinuationToken());
+
+                return this.feedContextClient.createDocumentChangeFeedQuery(this.feedContextClient.getContainerClient(), options)
+                    .take(1)
+                    .map(feedResponse -> {
+                        final String pkRangeIdSeparator = ":";
+                        final String segmentSeparator = "#";
+                        final String lsnPropertyName = "_lsn";
+                        final String tsPropertyName = "_ts";
+
+                        String sessionTokenLsn = feedResponse.getSessionToken();
+                        String parsedSessionToken = sessionTokenLsn.substring(sessionTokenLsn.indexOf(pkRangeIdSeparator));
+                        String[] segments = parsedSessionToken.split(segmentSeparator);
+                        String latestLsn = segments[0];
+
+                        if (segments.length >= 2) {
+                            // default to Global LSN
+                            latestLsn = segments[1];
+                        }
+
+                        ChangeFeedProcessorState changeFeedProcessorState = new ChangeFeedProcessorState()
+                            .setId(lease.getId())
+                            .setHostName(lease.getOwner())
+                            .setLeaseToken(lease.getLeaseToken())
+                            .setLastUpdate(ZonedDateTime.parse(lease.getTimestamp()).toInstant());
+
+
+                        // An empty list of documents returned means that we are current (zero lag)
+                        if (feedResponse.getResults() == null || feedResponse.getResults().size() == 0) {
```

Comment:
NIT: assign local variable for feedResponse.getResults() to avoid repetitive invocation of the property getter

Reply:
I'll keep it as such, it is easier to read the code. The property getter does not execute any particular complex operation, it just returns an internal member.

Reply:
also java compiler may inline these. so shouldn't be a perf hit.